Grbl_Esp32/src/Pin.h

```diff
@@ -0,0 +1,122 @@
+        static void IRAM_ATTR callback(void* ptr) { static_cast<ThisType*>(ptr)->*Callback(); }
+    };
+
+    // Helper for handling callbacks and mapping them to the proper class. This one is just meant
+    // for backward compatibility:
+    template <void (*Callback)()>
+    struct InterruptCallbackHelper2 {
+        static void IRAM_ATTR callback(void* /*ptr*/) { Callback(); }
+    };
+
+    // There are a few special indices:
+    //
+    // - 0-N = GPIO pins. These map 1:1 to native GPIO pins. N is 64 here.
+    // - 254 = undefined pin, maps to VoidPinDetail
+    // - 255 = fault pin (if you use it, it gives an error)
+    uint8_t _index;
+
+    inline Pin(uint8_t index) : _index(index) {}
+
+    static bool parse(String str, Pins::PinDetail*& detail);
+
+public:
+    using Capabilities = Pins::PinCapabilities;
+    using Attr         = Pins::PinAttributes;
+
+    static Pin UNDEFINED;
+    static Pin ERROR;
+
+    static const bool On  = true;
+    static const bool Off = false;
+
+    static Pin  create(const String& str);
+    static bool validate(const String& str);
+
+    inline Pin() : _index(254) {}  // Default to UNDEFINED
+
+    inline Pin(const Pin& o) = default;
+    inline Pin(Pin&& o)      = default;
+
+    inline Pin& operator=(const Pin& o) = default;
+    inline Pin& operator=(Pin&& o) = default;
+
+    // Some convenience operators:
+    inline bool operator==(const Pin& o) const { return _index == o._index; }
+    inline bool operator!=(const Pin& o) const { return _index != o._index; }
+
+    inline uint8_t getNative(Capabilities expectedBehavior) const {
+        auto detail = Pins::PinLookup::_instance.GetPin(_index);
+        Assert(detail->capabilities().has(expectedBehavior), "Requested pin does not have the expected behavior.");
+        return _index;
+    }
+
+    inline void write(bool value) const {
+        auto detail = Pins::PinLookup::_instance.GetPin(_index);
+        detail->write(value ? 1 : 0);
+    }
+
+    inline bool read() const {
+        auto detail = Pins::PinLookup::_instance.GetPin(_index);
+        return detail->read() != 0;
+    }
+
+    inline void setAttr(Attr attributes) const {
+        auto detail = Pins::PinLookup::_instance.GetPin(_index);
+        detail->setAttr(attributes);
+    }
+
+    inline void on() const { write(1); }
+    inline void off() const { write(0); }
+
+    // ISR handlers. Map methods on 'this' types.
+
+    template <typename ThisType, void (ThisType::*Callback)()>
+    void attachInterrupt(ThisType* arg, int mode) {
+        auto detail = Pins::PinLookup::_instance.GetPin(_index);
+        detail->attachInterrupt(InterruptCallbackHelper<ThisType, Callback>::callback, arg, mode);
+    }
+
+    // Backward compatibility ISR handler:
+    void attachInterrupt(void (*callback)(void*), int mode, void* arg = nullptr) const {
+        auto detail = Pins::PinLookup::_instance.GetPin(_index);
+        detail->attachInterrupt(callback, arg, mode);
```

Comment:
Changed this in System.cpp to remove the pullup to enable testing. using GPIO.34
    if (ControlResetPin->get() != Pin::UNDEFINED) {
        ControlResetPin->get().setAttr(Pin::Attr::Input); // | Pin::Attr::PullUp);
        ControlResetPin->get().attachInterrupt(isr_control_inputs, CHANGE);
    }
Needed to do this in Pin.h to avoid crashing upon attachInterrupt() call
    // Backward compatibility ISR handler:
    void attachInterrupt(void (*callback)(void*), int mode, void* arg = nullptr) const {
        auto detail = Pins::PinLookup::_instance.GetPin(_index);
    //    detail->attachInterrupt(callback, arg, mode);
    }

Reply:
Pullup is now handled differently in the latest branch, that shouldn't make a difference. Also, GPIO34 doesn't even have an internal pullup, so you cannot request it.

Unfortunately I still have to find the time to do ISR tests properly with endstops. What I do find weird tho is that `pio test -t test` works just fine with ISR's.